src/terminal/parser/InputStateMachineEngine.cpp

```diff
@@ -790,6 +858,113 @@ DWORD InputStateMachineEngine::_GetModifier(const size_t modifierParam) noexcept
+// Return Value:
+// true iff we were able to synthesize pdwButtonState
+bool InputStateMachineEngine::_UpdateSGRMouseButtonState(const wchar_t wch,
+                                                         const std::basic_string_view<size_t> parameters,
+                                                         DWORD& buttonState,
+                                                         DWORD& eventFlags)
+{
+    if (parameters.empty())
+    {
+        return false;
+    }
+
+    // Starting with the state from the last mouse event we received
+    buttonState = _mouseButtonState;
+    eventFlags = 0;
+    bool success = true;
+
+    // The first parameter of mouse events is encoded as the following two bytes:
+    // BBDM'MMBB
+    // Where each of the bits mean the following
+    //   BB__'__BB - which button was pressed/released
+    //   MMM - Control, Alt, Shift state (respectively)
+    //   D - flag signifying a drag event
+    // This retrieves the 2 MSBs and concatenates them to the 2 LSBs to create BBBB in binary
+    // This represents which button had a change in state
+    const auto buttonID = (til::at(parameters, 0) & 0x3) | ((til::at(parameters, 0) & 0xC0) >> 3);
+
+    // Step 1: Translate which button was affected
+    DWORD buttonFlag = 0;
+    switch (buttonID)
+    {
+    case CsiMouseButtonCodes::Left:
+        buttonFlag = FROM_LEFT_1ST_BUTTON_PRESSED;
+        break;
+    case CsiMouseButtonCodes::Right:
+        buttonFlag = RIGHTMOST_BUTTON_PRESSED;
+        break;
+    case CsiMouseButtonCodes::Middle:
+        buttonFlag = FROM_LEFT_2ND_BUTTON_PRESSED;
+        break;
+    default:
+        success = false;
+        break;
+    }
+
+    // Step 2: Decide whether to set or clear that button's bit
+    // NOTE: WI_SetFlag/WI_ClearFlag can't be used here because buttonFlag would have to be a compile-time constant
+    switch (static_cast<CsiActionCodes>(wch))
+    {
+    case CsiActionCodes::MouseDown:
+        // set flag
+        buttonState |= buttonFlag;
+        break;
+    case CsiActionCodes::MouseUp:
+        // clear flag
+        buttonState &= (~buttonFlag);
+        break;
+    default:
+        success = false;
+        break;
+    }
+
+    // Step 3: append the mouse wheel data to out param
+    switch (buttonID)
+    {
+    case CsiMouseButtonCodes::ScrollBack:
+    {
+        // set high word to proper scroll direction
+        // scroll intensity is assumed to be constant value
+        buttonState |= SCROLL_DELTA_BACKWARD;
+        eventFlags |= MOUSE_WHEELED;
+        break;
+    }
+    case CsiMouseButtonCodes::ScrollForward:
+    {
+        // set high word to proper scroll direction
+        // scroll intensity is assumed to be constant value
+        buttonState |= SCROLL_DELTA_FORWARD;
+        eventFlags |= MOUSE_WHEELED;
+        break;
+    }
+    default:
+        break;
+    }
+
+    // Step 4: check if mouse moved
+    if (WI_IsFlagSet(til::at(parameters, 0), CsiMouseModifierCodes::Drag))

```

Comment:
Instead of using `til::at(parameters, 0)` three times in here, it could just be pulled into a const local.